lib-ee/emqx_ee_bridge/src/gcp_pubsub/emqx_ee_bridge_gcp_pubsub_resource.erl

```diff
@@ -0,0 +1,599 @@
+                    #{instance_id => InstanceId, reason => Error}
+                ),
+                ?SLOG(error, #{
+                    msg => "failed_to_start_gcp_pubsub_jwt_worker",
+                    instance_id => InstanceId,
+                    pubsub_topic => PubSubTopic,
+                    reason => Error
+                }),
+                _ = emqx_connector_jwt_sup:ensure_worker_deleted(JWTWorkerId),
+                throw(failed_to_start_jwt_worker)
+        end,
+    MRef = monitor(process, Worker),
+    Ref = emqx_connector_jwt_worker:ensure_jwt(Worker),
+
+    %% to ensure that this resource and its actions will be ready to
+    %% serve when started, we must ensure that the first JWT has been
+    %% produced by the worker.
+    receive
+        {Ref, token_created} ->
+            ?tp(gcp_pubsub_bridge_jwt_created, #{resource_id => InstanceId}),
+            demonitor(MRef, [flush]),
+            ok;
+        {'DOWN', MRef, process, Worker, Reason} ->
+            ?tp(
+                gcp_pubsub_bridge_jwt_worker_failed_to_start,
+                #{
+                    resource_id => InstanceId,
+                    reason => Reason
+                }
+            ),
+            ?SLOG(error, #{
+                msg => "gcp_pubsub_bridge_jwt_worker_failed_to_start",
+                connector => InstanceId,
+                reason => Reason
+            }),
+            _ = emqx_connector_jwt_sup:ensure_worker_deleted(JWTWorkerId),
+            throw(failed_to_start_jwt_worker)
+    after 10_000 ->
+        ?tp(gcp_pubsub_bridge_jwt_timeout, #{resource_id => InstanceId}),
+        ?SLOG(warning, #{
+            msg => "gcp_pubsub_bridge_jwt_timeout",
+            connector => InstanceId
+        }),
+        demonitor(MRef, [flush]),
+        _ = emqx_connector_jwt_sup:ensure_worker_deleted(JWTWorkerId),
+        throw(timeout_creating_jwt)
+    end,
+    #{
+        jwt_worker_id => JWTWorkerId,
+        project_id => ProjectId
+    }.
+
+-spec encode_payload(state(), Selected :: map()) -> #{data := binary()}.
+encode_payload(_State = #{payload_template := PayloadTemplate}, Selected) ->
+    Interpolated =
+        case PayloadTemplate of
+            [] -> emqx_json:encode(Selected);
+            _ -> emqx_plugin_libs_rule:proc_tmpl(PayloadTemplate, Selected)
+        end,
+    #{data => base64:encode(Interpolated)}.
+
+-spec to_pubsub_request([#{data := binary()}]) -> binary().
+to_pubsub_request(Payloads) ->
+    emqx_json:encode(#{messages => Payloads}).
+
+-spec publish_path(state()) -> binary().
+publish_path(
+    _State = #{
+        project_id := ProjectId,
+        pubsub_topic := PubSubTopic
+    }
+) ->
+    <<"/v1/projects/", ProjectId/binary, "/topics/", PubSubTopic/binary, ":publish">>.
+
+-spec get_jwt_authorization_header(resource_id()) -> [{binary(), binary()}].
+get_jwt_authorization_header(InstanceId) ->
+    case emqx_connector_jwt:lookup_jwt(?JWT_TABLE, InstanceId) of
```

Comment:
NIT: Just an idea, since the intended usage is async, then for simplicity, couldn't we just lazily generate and cache JWT here?

Reply:
I wanted to avoid duplicate work (like when a pool of workers simultaneously notice they need to refresh/issue a new JWT) and also to refresh it some time before expiration (although this second point might still be done lazily by checking the expiration time :thinking: ).